### PR TITLE
Backports for 2.7.3

### PR DESCRIPTION
**Pull request by hasufell**
@bgamari 

Backports:

* https://github.com/haskell/unix/pull/229
* https://github.com/haskell/unix/pull/228
* https://github.com/haskell/unix/pull/207
* https://github.com/haskell/unix/pull/144
* https://github.com/haskell/unix/pull/122
* https://github.com/haskell/unix/pull/119

### CODE REVIEW

File: System/Posix/ByteString/FilePath.hsc [hunk 1]
```diff
@@ -125,3 +129,9 @@ throwErrnoPathIfMinus1 = throwErrnoPathIf (== -1)
 --
 throwErrnoPathIfMinus1_ :: (Eq a, Num a) => String -> RawFilePath -> IO a -> IO ()
 throwErrnoPathIfMinus1_  = throwErrnoPathIf_ (== -1)
+
+-- | as 'throwErrnoTwoPathsIfMinus1_', but exceptions include two paths when appropriate.
+--
+throwErrnoTwoPathsIfMinus1_ :: (Eq a, Num a) => String -> RawFilePath -> RawFilePath -> IO a -> IO ()
+throwErrnoTwoPathsIfMinus1_  loc path1 path2 =
+    throwErrnoIfMinus1_ (loc <> " '" <> BC.unpack path1 <> "' to '" <> BC.unpack path2 <> "'")
```
Comment (Bodigrim) [inline System/Posix/ByteString/FilePath.hsc:137]: This is strictly speaking a breaking change: someone could reasonably catch the error message and analyse it.

File: System/Posix/Env/ByteString.hsc [hunk 2]
```diff
@@ -100,6 +101,18 @@ getEnvironment = do
       | BC.head y == '=' = (x,B.tail y)
       | otherwise       = error $ "getEnvironment: insane variable " ++ BC.unpack x
 
+-- |'setEnvironment' resets the entire environment to the given list of
+-- @(key,value)@ pairs.
+--
+-- @since 2.8.0.0
```
Comment (Bodigrim) [inline System/Posix/Env/ByteString.hsc:107]: Should not be 2.8.0.0.

**Review by Bodigrim — COMMENTED** [review]
To be honest I'd recommend to leave `unix-2.7` in peace. At the very least, this is not `2.7.2.3` but rather `2.7.3`.

File: System/Posix/ByteString/FilePath.hsc (hunk 1 above)
Comment (hasufell) [reply System/Posix/ByteString/FilePath.hsc:137]: Well, not per PVP, because the API is defined in terms of types only, not values.

Comment (hasufell) [reply System/Posix/ByteString/FilePath.hsc:137]: Sure, I can remove this patchset, since it's not strictly speaking a bugfix.

Comment (hasufell) [reply System/Posix/ByteString/FilePath.hsc:137]: > technically allows me to replace every function call with undefined.

You're giving me ideas :smiling_imp:

**Comment by Bodigrim** [comment]
CI does not check for GHC < 8.2, so I suggest we bump the lower bound for `base` to match. Unless we have a volunteer to test GHC 7.4 ;)

**Comment by Bodigrim** [comment]
Ah, there is no CI at all in `2.7` branch. Sigh.

**Comment by hasufell** [comment]
> Ah, there is no CI at all in `2.7` branch. Sigh.

Added.

**Review by Bodigrim — APPROVED** [review]

**Review by hs-viktor — APPROVED** [review]
The code changes LGTM.  Note I have not closely examined all the changes in the test tree.  I don't know why they're needed.  If the rest of the team is cool with that, go ahead and merge...

**Pull request merged**